File: vito/core/src/main/java/com/facebook/fresco/vito/core/impl/ImagePipelineUtilsImpl.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl;

import android.net.Uri;
import com.facebook.fresco.vito.core.ImagePipelineUtils;
import com.facebook.fresco.vito.options.DecodedImageOptions;
import com.facebook.fresco.vito.options.EncodedImageOptions;
import com.facebook.imagepipeline.common.ImageDecodeOptions;
import com.facebook.imagepipeline.common.Priority;
import com.facebook.imagepipeline.common.ResizeOptions;
import com.facebook.imagepipeline.common.RotationOptions;
import com.facebook.imagepipeline.request.ImageRequest;
import com.facebook.imagepipeline.request.ImageRequestBuilder;
import com.facebook.imagepipeline.request.Postprocessor;
import com.facebook.infer.annotation.Nullsafe;
import javax.annotation.Nullable;

/**
 * Utility methods to create {@link ImageRequest}s for {@link
 * com.facebook.fresco.vito.options.ImageOptions}.
 */
@Nullsafe(Nullsafe.Mode.LOCAL)
public class ImagePipelineUtilsImpl implements ImagePipelineUtils {

  public interface CircularBitmapRounding {

    ImageDecodeOptions getDecodeOptions(boolean antiAliased);
  }

  public interface ImageDecodeOptionsProvider {

    @Nullable
    ImageDecodeOptions create(
        ImageRequestBuilder imageRequestBuilder, DecodedImageOptions imageOptions);
  }

  private final ImageDecodeOptionsProvider mImageDecodeOptionsProvider;

  public ImagePipelineUtilsImpl(ImageDecodeOptionsProvider imageDecodeOptionsProvider) {
    mImageDecodeOptionsProvider = imageDecodeOptionsProvider;
  }

  @Override
  @Nullable
  public ImageRequest buildImageRequest(@Nullable Uri uri, DecodedImageOptions imageOptions) {
    if (uri == null) {
      return null;
    }
    final ImageRequestBuilder imageRequestBuilder =
        createEncodedImageRequestBuilder(uri, imageOptions);
    ImageRequestBuilder builder =
        createDecodedImageRequestBuilder(imageRequestBuilder, imageOptions);
    return builder != null ? builder.build() : null;
  }

  @Override
  @Nullable
  public ImageRequest wrapDecodedImageRequest(
      ImageRequest imageRequest, DecodedImageOptions imageOptions) {
    ImageRequestBuilder builder =
        createDecodedImageRequestBuilder(
            createEncodedImageRequestBuilder(imageRequest, imageOptions), imageOptions);
    return builder != null ? builder.build() : null;
  }

  @Override
  @Nullable
  public ImageRequest buildEncodedImageRequest(
      @Nullable Uri uri, EncodedImageOptions imageOptions) {
    ImageRequestBuilder builder = createEncodedImageRequestBuilder(uri, imageOptions);
    return builder != null ? builder.build() : null;
  }

  @Nullable
  protected ImageRequestBuilder createDecodedImageRequestBuilder(
      @Nullable ImageRequestBuilder imageRequestBuilder, DecodedImageOptions imageOptions) {
    if (imageRequestBuilder == null) {
      return null;
    }

    ResizeOptions resizeOptions = imageOptions.getResizeOptions();
    if (resizeOptions != null) {
      imageRequestBuilder.setResizeOptions(resizeOptions);
    }

    RotationOptions rotationOptions = imageOptions.getRotationOptions();
    if (rotationOptions != null) {
      imageRequestBuilder.setRotationOptions(rotationOptions);
    }

    ImageDecodeOptions imageDecodeOptions =
        mImageDecodeOptionsProvider.create(imageRequestBuilder, imageOptions);
    if (imageDecodeOptions != null) {
      imageRequestBuilder.setImageDecodeOptions(imageDecodeOptions);
    }

    imageRequestBuilder.setLocalThumbnailPreviewsEnabled(
        imageOptions.areLocalThumbnailPreviewsEnabled());

    Postprocessor postprocessor = imageOptions.getPostprocessor();
    if (postprocessor != null) {
      imageRequestBuilder.setPostprocessor(postprocessor);
    }

    if (imageOptions.isProgressiveDecodingEnabled() != null) {
      imageRequestBuilder.setProgressiveRenderingEnabled(
          imageOptions.isProgressiveDecodingEnabled());
    }

    return imageRequestBuilder;
  }

  @Nullable
  protected ImageRequestBuilder createEncodedImageRequestBuilder(
      @Nullable Uri uri, EncodedImageOptions imageOptions) {
    if (uri == null) {
      return null;
    }
    ImageRequestBuilder builder = ImageRequestBuilder.newBuilderWithSource(uri);
    maybeSetRequestPriority(builder, imageOptions.getPriority());
    return builder;
  }

  @Nullable
  protected ImageRequestBuilder createEncodedImageRequestBuilder(
      ImageRequest imageRequest, EncodedImageOptions imageOptions) {
    if (imageRequest == null) {
      return null;
    }
    ImageRequestBuilder builder = ImageRequestBuilder.fromRequest(imageRequest);
    maybeSetRequestPriority(builder, imageOptions.getPriority());
    return builder;
  }

  private static void maybeSetRequestPriority(
      ImageRequestBuilder builder, @Nullable Priority priority) {
    if (priority != null) {
      builder.setRequestPriority(priority);
    }
  }
}

File: vito/core-java-impl/src/main/java/com/facebook/fresco/vito/core/impl/ImagePipelineUtilsImpl.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl

import android.net.Uri
import com.facebook.fresco.vito.core.ImagePipelineUtils
import com.facebook.fresco.vito.options.DecodedImageOptions
import com.facebook.fresco.vito.options.EncodedImageOptions
import com.facebook.imagepipeline.common.ImageDecodeOptions
import com.facebook.imagepipeline.common.Priority
import com.facebook.imagepipeline.request.ImageRequest
import com.facebook.imagepipeline.request.ImageRequestBuilder

/**
 * Utility methods to create [ImageRequest]s for [com.facebook.fresco.vito.options.ImageOptions].
 */
class ImagePipelineUtilsImpl(private val imageDecodeOptionsProvider: ImageDecodeOptionsProvider) :
    ImagePipelineUtils {

  fun interface CircularBitmapRounding {
    fun getDecodeOptions(antiAliased: Boolean): ImageDecodeOptions?
  }

  fun interface ImageDecodeOptionsProvider {
    fun create(
        imageRequestBuilder: ImageRequestBuilder,
        imageOptions: DecodedImageOptions
    ): ImageDecodeOptions?
  }

  override fun buildImageRequest(uri: Uri?, imageOptions: DecodedImageOptions): ImageRequest? {
    if (uri == null) {
      return null
    }
    return createDecodedImageRequestBuilder(
            createEncodedImageRequestBuilder(uri, imageOptions), imageOptions)
        ?.build()
  }

  override fun wrapDecodedImageRequest(
      originalRequest: ImageRequest,
      imageOptions: DecodedImageOptions
  ): ImageRequest? =
      createDecodedImageRequestBuilder(
              createEncodedImageRequestBuilder(originalRequest, imageOptions), imageOptions)
          ?.build()

  override fun buildEncodedImageRequest(
      uri: Uri?,
      imageOptions: EncodedImageOptions
  ): ImageRequest? = createEncodedImageRequestBuilder(uri, imageOptions)?.build()

  protected fun createDecodedImageRequestBuilder(
      imageRequestBuilder: ImageRequestBuilder?,
      imageOptions: DecodedImageOptions
  ): ImageRequestBuilder? =
      imageRequestBuilder?.apply {
        imageOptions.resizeOptions?.let { resizeOptions = it }
        imageOptions.downsampleOverride?.let { downsampleOverride = it }
        imageOptions.rotationOptions?.let { rotationOptions = it }
        imageDecodeOptionsProvider.create(imageRequestBuilder, imageOptions)?.let {
          imageDecodeOptions = it
        }
        isLocalThumbnailPreviewsEnabled = imageOptions.areLocalThumbnailPreviewsEnabled()
        loadThumbnailOnly = imageOptions.loadThumbnailOnly
        imageOptions.postprocessor?.let { postprocessor = it }
        imageOptions.isProgressiveDecodingEnabled?.let { isProgressiveRenderingEnabled = it }
      }

  protected fun createEncodedImageRequestBuilder(
      uri: Uri?,
      imageOptions: EncodedImageOptions
  ): ImageRequestBuilder? {
    if (uri == null) {
      return null
    }
    val builder = ImageRequestBuilder.newBuilderWithSource(uri)
    maybeSetRequestPriority(builder, imageOptions.priority)
    if (imageOptions.cacheChoice != null) {
      builder.cacheChoice = imageOptions.cacheChoice
    }
    if (imageOptions.diskCacheId != null) {
      builder.diskCacheId = imageOptions.diskCacheId
    }
    return builder
  }

  protected fun createEncodedImageRequestBuilder(
      imageRequest: ImageRequest?,
      imageOptions: EncodedImageOptions
  ): ImageRequestBuilder? {
    if (imageRequest == null) {
      return null
    }
    val builder = ImageRequestBuilder.fromRequest(imageRequest)
    maybeSetRequestPriority(builder, imageOptions.priority)
    return builder
  }

  companion object {
    private fun maybeSetRequestPriority(builder: ImageRequestBuilder, priority: Priority?) {
      if (priority != null) {
        builder.requestPriority = priority
      }
    }
  }
}
